magma/generator.py

```diff
@@ -55,3 +57,58 @@ def generate(*args, **kwargs):
     @classmethod
     def bind(cls, monitor):
         cls.bind_generators.append(monitor)
+
+
+def _make_key(*args, **kwargs):
+    _SECRET_KEY = "__magma_generator2_secret_key__"
+    dct = {f"{_SECRET_KEY}{i}": v for i, v in enumerate(args)}
+    dct.update(kwargs)
+    return ParamDict(dct)
+
+
+def _make_type(cls, *args, **kwargs):
+    dummy = type.__new__(cls, "", (), {})
+    name = cls.__name__
+    bases = (Circuit,)
+    dct = DefineCircuitKind.__prepare__(name, bases)
+    cls.__init__(dummy, *args, **kwargs)
+    dct.update(dict(dummy.__dict__))
+    return DefineCircuitKind.__new__(cls, name, bases, dct)
+
+
+class _Generator2Meta(type):
+    _cache = weakref.WeakValueDictionary()
+
+    def __new__(metacls, name, bases, dct):
+        bases = bases + (DefineCircuitKind,)
+        return type.__new__(metacls, name, bases, dct)
+
+    def __call__(*args, **kwargs):
+        cls = args[0]
+        args = args[1:]
+        if cls is Generator2:
+            try:
+                name, bases, dct = args
+            except ValueError:
+                raise Exception("Can not initialize base Generator class")
+            assert not kwargs
+            return type.__new__(cls, name, bases, dct)
+        cache = not getattr(cls, "_no_cache_", False)
```

Comment:
seems a bit strange why not do `getattr(cls, "_cache_", True)`

Reply:
agreed